### PR TITLE
`Programming exercise`: Use the git-diff to show line stats 

### PR DESCRIPTION
**Pull request by Hialus**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Since we now generate the git-diff report for every exercise it makes sense to use this information in other ways.
One way to utilize this is to show the amount of work that the students have to do for an exercise.

### Description

The amount of work can be roughly described by the number of lines the students have to add/change. The simplest way to show this is a line stat view that shows the added and removed lines from the git-diff. This line stat view was inspired by the one GitHub uses.
The line-stat view can be seen in 3 places:
1. At the bottom of the detail view (if the git-diff report exists)
3. At the top right of the git-diff report modal
4. At the top right of each git-diff entry in the modal
    - This one is a bit temporary, as it will be reworked when we improve the diff view

One thing to note here is that 2 and 3 represent filtered line stats meaning that they filter out empty lines, as these don't really represent any amount of work. 1 on the other hand represents an unfiltered line stat, as we do not have access to the full git-diff report (with the actual code) here.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise

If you encounter an internal server error when creating new exercises see #4893 

1. Log in to Artemis
5. Go to the exercise detail page
7. Open the git diff modal
8. Check that the filtered line counts are correct
9. Close the modal
6. Look at the bottom if the unfiltered line count is correct

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| New client code | 100% | 100% |

### Screenshots

The git-diff modal: 
grafik
The detail page:
grafik

### CODE REVIEW

File: src/main/webapp/app/exercises/programming/hestia/git-diff-report/full-git-diff-entry.component.ts [hunk 1]
```diff
@@ -13,12 +13,16 @@ export class FullGitDiffEntryComponent implements OnInit {
...
+    addedLineCount: number;
+    removedLineCount: number;
 
     constructor() {}
 
     ngOnInit(): void {
         this.setupEditor(this.editorPrevious, this.diffEntry.previousLine, this.diffEntry.previousCode ?? '', 'rgba(248, 81, 73, 0.5)');
         this.setupEditor(this.editorNow, this.diffEntry.line, this.diffEntry.code ?? '', 'rgba(63, 185, 80, 0.5)');
+        this.addedLineCount = this.diffEntry.code?.split('\n').filter((line) => line !== undefined && line.trim().length !== 0).length ?? 0;
+        this.removedLineCount = this.diffEntry.previousCode?.split('\n').filter((line) => line !== undefined && line.trim().length !== 0).length ?? 0;
```
Comment (ole-ve) [inline src/main/webapp/app/exercises/programming/hestia/git-diff-report/full-git-diff-entry.component.ts:25]:
```suggestion
        this.addedLineCount = this.diffEntry.code?.split('\n').filter((line) => line && line.trim().length !== 0).length ?? 0;
        this.removedLineCount = this.diffEntry.previousCode?.split('\n').filter((line) => line && line.trim().length !== 0).length ?? 0;
```

File: src/main/webapp/app/exercises/programming/hestia/git-diff-report/full-git-diff-report.component.ts [hunk 2]
```diff
@@ -26,5 +28,8 @@ export class FullGitDiffReportComponent implements OnInit {
             }
             return (a.line ?? a.previousLine ?? 0) - (b.line ?? b.previousLine ?? 0);
         });
+
+        this.addedLineCount = this.entries.flatMap((entry) => entry.code?.split('\n')).filter((line) => line !== undefined && line.trim().length !== 0).length;
+        this.removedLineCount = this.entries.flatMap((entry) => entry.previousCode?.split('\n')).filter((line) => line !== undefined && line.trim().length !== 0).length;
```
Comment (ole-ve) [inline src/main/webapp/app/exercises/programming/hestia/git-diff-report/full-git-diff-report.component.ts:33]: Same here

File: src/main/webapp/app/exercises/programming/hestia/git-diff-report/git-diff-line-stat.component.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    ngOnInit(): void {
+        if (this.addedLineCount === 0) {
+            this.addedSquareCount = 0;
+            this.removedSquareCount = 5;
+        } else if (this.removedLineCount === 0) {
+            this.addedSquareCount = 5;
+            this.removedSquareCount = 0;
+        } else {
+            const totalLineCount = this.addedLineCount + this.removedLineCount;
+            this.addedSquareCount = Math.round(Math.max(1, Math.min(4, (this.addedLineCount / totalLineCount) * 5)));
```
Comment (ole-ve) [inline src/main/webapp/app/exercises/programming/hestia/git-diff-report/git-diff-line-stat.component.ts:27]: I think just a small comment on what this calculation is about would be useful

File: src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts [hunk 4]
```diff
@@ -176,7 +178,21 @@ export class ProgrammingExerciseDetailComponent implements OnInit, OnDestroy {
                 }
             });
 
-            this.programmingExerciseService.getDiffReport(programmingExercise.id).subscribe((gitDiffReport) => (this.programmingExercise.gitDiffReport = gitDiffReport));
+            this.programmingExerciseService.getDiffReport(programmingExercise.id).subscribe((gitDiffReport) => {
+                this.programmingExercise.gitDiffReport = gitDiffReport;
+                if (gitDiffReport !== undefined) {
```
Comment (ole-ve) [inline src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts:183]:
```suggestion
                if (gitDiffReport) {
```

File: src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts [hunk 5]
```diff
@@ -176,7 +178,21 @@ export class ProgrammingExerciseDetailComponent implements OnInit, OnDestroy {
...
+                    this.addedLineCount = gitDiffReport.entries
+                        .map((entry) => entry.lineCount)
+                        .filter((lineCount) => lineCount !== undefined)
```
Comment (ole-ve) [inline src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts:186]:
```suggestion
                        .filter((lineCount) => lineCount)
```

File: src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts [hunk 6]
```diff
@@ -176,7 +178,21 @@ export class ProgrammingExerciseDetailComponent implements OnInit, OnDestroy {
...
+                        .map((lineCount) => lineCount!)
+                        .reduce((lineCount1, lineCount2) => lineCount1 + lineCount2, 0);
+                    this.removedLineCount = gitDiffReport.entries
+                        .map((entry) => entry.previousLineCount)
+                        .filter((lineCount) => lineCount !== undefined)
```
Comment (ole-ve) [inline src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.ts:191]:
```suggestion
                        .filter((lineCount) => lineCount)
```

File: src/main/webapp/i18n/de/programmingExercise.json [hunk 7]
```diff
@@ -387,7 +387,11 @@
                 "title": "Template-Lösungs-Diff",
                 "template": "Template",
                 "solution": "Lösung",
-                "404": "Template-Lösungs-Diff wurde noch nicht generiert. Bitte pusche etwas in die Template- oder Lösungs-Repositories um ihn zu erstellen."
+                "404": "Template-Lösungs-Diff wurde noch nicht generiert. Bitte pusche etwas in die Template- oder Lösungs-Repositories um ihn zu erstellen.",
```
Comment (ole-ve) [inline src/main/webapp/i18n/de/programmingExercise.json:390]:
```suggestion
                "404": "Template-Lösungs-Diff wurde noch nicht generiert. Bitte pusche etwas in die Template- oder Lösungs-Repositories, um ihn zu erstellen.",
```

File: src/main/webapp/i18n/de/programmingExercise.json [hunk 8]
```diff
@@ -387,7 +387,11 @@
...
+                "lineStatLabel": "Zeilen die zwischen Vorlage und Lösung hinzugefügt/entfernt wurden",
```
Comment (ole-ve) [inline src/main/webapp/i18n/de/programmingExercise.json:391]:
```suggestion
                "lineStatLabel": "Zeilen, die zwischen Vorlage und Lösung hinzugefügt/entfernt wurden",
```

File: src/test/javascript/spec/component/hestia/git-diff-report/git-diff-line-stat.component.spec.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        }).compileComponents();
+        fixture = TestBed.createComponent(GitDiffLineStatComponent);
+        comp = fixture.componentInstance;
+    });
+
+    afterEach(() => {
+        jest.restoreAllMocks();
+    });
+
+    it('Should show 5-0 boxes', () => {
```
Comment (ole-ve) [inline src/test/javascript/spec/component/hestia/git-diff-report/git-diff-line-stat.component.spec.ts:23]: I think the tests could well be extracted into an own method

**Review by ole-ve — COMMENTED** [review]
I have added some inline comments. Manual testing will follow.

File: src/test/javascript/spec/component/hestia/git-diff-report/git-diff-line-stat.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+            declarations: [GitDiffLineStatComponent],
+            providers: [],
+        }).compileComponents();
+        fixture = TestBed.createComponent(GitDiffLineStatComponent);
+        comp = fixture.componentInstance;
+    });
+
+    afterEach(() => {
+        jest.restoreAllMocks();
+    });
```
Comment (ole-ve) [inline src/test/javascript/spec/component/hestia/git-diff-report/git-diff-line-stat.component.spec.ts:21]: This could be removed since there are no mocks

**Review by ole-ve — APPROVED** [review]
Approve Code 👍
(Approve for manual testing is still valid since code has only changed minor)

**Review by canberkanar — APPROVED** [review]
Tested on ts2. Empty lines are omitted and the box calculations work according to the in-code comment. Code also looks good.

**Pull request merged**